### PR TITLE
Add proof harnesses for the list_utils module

### PR DESCRIPTION
**Pull request by danielsn**
*Issue #, if available:*
N/A.

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    const struct aws_array_list old_dest = *dest;
+    const struct aws_array_list old_src  = *src;
+
+    if (aws_cryptosdk_edk_list_copy_all(can_fail_allocator(), dest, src) == AWS_OP_SUCCESS) {  // Finished in
+        assert(src->length == old_src.length);
+        assert(dest->length == old_dest.length + old_src.length);
+    } else {
+        assert(src->length == old_src.length);
+        assert(dest->length == old_dest.length);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c:89]: For now, it's ok, but on a second round we could add a check to make sure if the copy is successful the bytes match. If it's not successful, the bytes should match with the `old_*`states. Shall we create an issue on GitHub as "Add post-conditions to ESDK proofs"?

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    assert(aws_array_list_is_valid(src));
+    assert(aws_array_list_is_valid(dest));
+}
+
+#include <aws/common/error.inl>
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c:95]: why here?

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_copy_all/Makefile [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+#########
+# if Makefile.local exists, use it. This provides a way to override the defaults
+sinclude ../Makefile.local
+#otherwise, use the default values
+include ../Makefile.local_default
+
+#########
+# Local vars
+NUM_ELEMS = 2
+ITEM_SIZE = 40
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_copy_all/Makefile:23]: Apply to this file all the comments I left on the previous Makefile.

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_copy_all/aws_cryptosdk_keyring_trace_copy_all_harness.c [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+bool aws_array_list_is_valid(const struct aws_array_list *AWS_RESTRICT list) {
+    if (!list) {
+        return false;
+    }
+
+    bool data_is_valid =
+        ((list->current_size == 0 && list->data == NULL) || AWS_MEM_IS_WRITABLE(list->data, list->current_size));
+    bool item_size_is_valid = (list->item_size != 0);
+    return data_is_valid && item_size_is_valid;
+}
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_copy_all/aws_cryptosdk_keyring_trace_copy_all_harness.c:34]: Apply to this file all the comments I left on the previous harness.

File: .cbmc-batch/jobs/aws_cryptosdk_transfer_list/Makefile [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+#########
+# if Makefile.local exists, use it. This provides a way to override the defaults
+sinclude ../Makefile.local
+#otherwise, use the default values
+include ../Makefile.local_default
+
+#########
+# Local vars
+NUM_ELEMS = 2
+ITEM_SIZE = 4
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_transfer_list/Makefile:23]: Apply to this file all the comments I left on the previous Makefile.

File: .cbmc-batch/jobs/aws_cryptosdk_transfer_list/Makefile [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+#2,4 = 11 min
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_transfer_list/Makefile:25]: Expand this comment and put it along with the local variables

File: .cbmc-batch/jobs/aws_cryptosdk_transfer_list/Makefile [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+#########
+ABSTRACTIONS += $(SRCDIR)/c-common-helper-stubs/error.c
+#ABSTRACTIONS += $(SRCDIR)/c-common-helper-stubs/memcmp_override_no_op.c
+#ABSTRACTIONS += $(SRCDIR)/c-common-helper-stubs/aws_array_list_defined_type.c
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_transfer_list/Makefile:30]: Remove it if is not necessary.

File: .cbmc-batch/jobs/aws_cryptosdk_transfer_list/aws_cryptosdk_transfer_list_harness.c [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+bool aws_array_list_is_valid(const struct aws_array_list *AWS_RESTRICT list) {
+    if (!list) {
+        return false;
+    }
+
+    bool data_is_valid =
+        ((list->current_size == 0 && list->data == NULL) || AWS_MEM_IS_WRITABLE(list->data, list->current_size));
+    bool item_size_is_valid = (list->item_size != 0);
+    return data_is_valid && item_size_is_valid;
+}
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_transfer_list/aws_cryptosdk_transfer_list_harness.c:34]: Apply to this file all the comments I left on the previous harness.

File: .cbmc-batch/jobs/aws_cryptosdk_transfer_list/aws_cryptosdk_transfer_list_harness.c [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+bool aws_array_list_is_valid_deep(const struct aws_array_list *AWS_RESTRICT list) {
+    if (!list) {
+        return false;
+    }
+    size_t required_size        = list->length * list->item_size;
+    bool required_size_is_valid = true;
+    //        (aws_mul_size_checked(list->length, list->item_size, &required_size) == AWS_OP_SUCCESS);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_transfer_list/aws_cryptosdk_transfer_list_harness.c:42]: Does it blows up CBMC? Why comment it? Should it be removed from it?

File: .cbmc-batch/jobs/list_copy_all/Makefile [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+sinclude ../Makefile.local
+#otherwise, use the default values
+include ../Makefile.local_default
+
+#########
+# Local vars
+NUM_ELEMS = 2
+ITEM_SIZE = 4
+
+#2,4 = 11 min
```
Comment (feliperodri) [inline .cbmc-batch/jobs/list_copy_all/Makefile:25]: Apply to this file all the comments I left on the previous Makefile.

File: .cbmc-batch/jobs/list_copy_all/list_copy_all_harness.c [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+bool aws_array_list_is_valid(const struct aws_array_list *AWS_RESTRICT list) {
+    if (!list) {
+        return false;
+    }
+
+    bool data_is_valid =
+        ((list->current_size == 0 && list->data == NULL) || AWS_MEM_IS_WRITABLE(list->data, list->current_size));
+    bool item_size_is_valid = (list->item_size != 0);
+    return data_is_valid && item_size_is_valid;
+}
```
Comment (feliperodri) [inline .cbmc-batch/jobs/list_copy_all/list_copy_all_harness.c:34]: Apply to this file all the comments I left on the previous harness.

File: source/list_utils.c [hunk 12]
```diff
@@ -35,13 +40,17 @@ int aws_cryptosdk_transfer_list(struct aws_array_list *dest, struct aws_array_li
 typedef int (*clone_item_fn)(struct aws_allocator *, void *, const void *);
 typedef void (*clean_up_item_fn)(void *);
 
-static int list_copy_all(
+/*static*/ int list_copy_all(
```
Comment (feliperodri) [inline source/list_utils.c:43]: Why? If we are commenting out, why don't just remove it?

File: source/list_utils.c [hunk 13]
```diff
@@ -35,13 +40,17 @@ int aws_cryptosdk_transfer_list(struct aws_array_list *dest, struct aws_array_li
...
     struct aws_allocator *alloc,
     struct aws_array_list *dest,
     const struct aws_array_list *src,
     clone_item_fn cloner,
     clean_up_item_fn cleaner) {
-    assert(dest->item_size == src->item_size);
+    AWS_ERROR_PRECONDITION(aws_allocator_is_valid(alloc));
```
Comment (feliperodri) [inline source/list_utils.c:49]: (picky) you could move this validity check together with the others and leave `src != dest` first.

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/Makefile [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+#########
+# if Makefile.local exists, use it. This provides a way to override the defaults
+sinclude ../Makefile.local
+#otherwise, use the default values
+include ../Makefile.local_default
+
+#########
+# Local vars
+NUM_ELEMS = 2
+ITEM_SIZE = 40
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/Makefile:23]: Done

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/Makefile [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+DEPENDENCIES += $(SRCDIR)/c-common-src/array_list.c
+DEPENDENCIES += $(SRCDIR)/c-common-src/byte_buf.c
+DEPENDENCIES += $(SRCDIR)/c-common-src/common.c
+DEPENDENCIES += $(SRCDIR)/c-common-src/math.c
+DEPENDENCIES += $(SRCDIR)/c-enc-sdk-src/list_utils.c
+DEPENDENCIES += $(SRCDIR)/helper-src/make_common_data_structures.c
+
+ENTRY = aws_cryptosdk_edk_list_copy_all_harness
+
+REMOVE_FUNCTION_BODIES += --remove-function-body aws_array_list_is_valid
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/Makefile:48]: Done

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/Makefile [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+addone = $(shell echo $$(( $(1) + 1)))
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/Makefile:49]: Done

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+bool aws_array_list_is_valid(const struct aws_array_list *AWS_RESTRICT list) {
+    if (!list) {
+        return false;
+    }
+
+    bool data_is_valid =
+        ((list->current_size == 0 && list->data == NULL) || AWS_MEM_IS_WRITABLE(list->data, list->current_size));
+    bool item_size_is_valid = (list->item_size != 0);
+    return data_is_valid && item_size_is_valid;
+}
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c:34]: Done.  Tracking issue created

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+        return false;
+    }
+    size_t required_size        = list->length * list->item_size;
+    bool required_size_is_valid = true;
+    bool current_size_is_valid  = (list->current_size >= required_size);
+    bool data_is_valid =
+        ((list->current_size == 0 && list->data == NULL) || AWS_MEM_IS_WRITABLE(list->data, list->current_size));
+    bool item_size_is_valid = (list->item_size != 0);
+    return required_size_is_valid && current_size_is_valid && data_is_valid && item_size_is_valid;
+}
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c:47]: Done

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+
+// allocator, dest, src
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c:49]: Yes

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+    assert(AWS_MEM_IS_WRITABLE(dest, g_item_size));
+    uint8_t *d = (uint8_t *)dest;
+    *d         = 0xab;
+    return nondet_int();
+}
+
+void aws_cryptosdk_edk_clean_up(void *p) {
+    uint8_t *d = (uint8_t *)p;
+    assert(*d == 0xab);
+}
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c:66]: Comment added.

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+    struct aws_array_list *src = can_fail_malloc(sizeof(*src));
+    __CPROVER_assume(src != NULL);
+    __CPROVER_assume(aws_array_list_is_bounded(src, NUM_ELEMS, sizeof(struct aws_cryptosdk_edk)));
+    ensure_array_list_has_allocated_data_member(src);
+    __CPROVER_assume(aws_array_list_is_valid_deep(src));
+
+    const struct aws_array_list old_dest = *dest;
+    const struct aws_array_list old_src  = *src;
+
+    if (aws_cryptosdk_edk_list_copy_all(can_fail_allocator(), dest, src) == AWS_OP_SUCCESS) {  // Finished in
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c:84]: Removed.

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c (hunk 1 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c:89]: https://github.com/aws/aws-encryption-sdk-c/issues/518

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c (hunk 2 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/aws_cryptosdk_edk_list_copy_all_harness.c:95]: Fixed.

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_copy_all/Makefile (hunk 3 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_copy_all/Makefile:23]: Done

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_copy_all/aws_cryptosdk_keyring_trace_copy_all_harness.c (hunk 4 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_copy_all/aws_cryptosdk_keyring_trace_copy_all_harness.c:34]: Done

File: .cbmc-batch/jobs/aws_cryptosdk_transfer_list/Makefile (hunk 5 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_transfer_list/Makefile:23]: Done

File: .cbmc-batch/jobs/aws_cryptosdk_transfer_list/Makefile (hunk 6 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_transfer_list/Makefile:25]: done

File: .cbmc-batch/jobs/aws_cryptosdk_transfer_list/Makefile (hunk 7 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_transfer_list/Makefile:30]: done

File: .cbmc-batch/jobs/aws_cryptosdk_transfer_list/aws_cryptosdk_transfer_list_harness.c (hunk 8 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_transfer_list/aws_cryptosdk_transfer_list_harness.c:34]: Done

File: .cbmc-batch/jobs/list_copy_all/list_copy_all_harness.c (hunk 11 above)
Comment (danielsn) [reply .cbmc-batch/jobs/list_copy_all/list_copy_all_harness.c:34]: done

File: source/list_utils.c (hunk 13 above)
Comment (danielsn) [reply source/list_utils.c:49]: I liked the idea of checking in the same order as the function params.

File: .cbmc-batch/jobs/list_copy_all/Makefile (hunk 10 above)
Comment (danielsn) [reply .cbmc-batch/jobs/list_copy_all/Makefile]: done

File: source/list_utils.c (hunk 12 above)
Comment (danielsn) [reply source/list_utils.c]: changed the makefiles so we can use the static version

File: .cbmc-batch/jobs/aws_cryptosdk_transfer_list/aws_cryptosdk_transfer_list_harness.c (hunk 9 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_transfer_list/aws_cryptosdk_transfer_list_harness.c]: fixed

File: .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/Makefile [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+DEPENDENCIES += $(SRCDIR)/c-common-src/math.c
+DEPENDENCIES += $(SRCDIR)/c-enc-sdk-src/list_utils.c
+DEPENDENCIES += $(SRCDIR)/helper-src/make_common_data_structures.c
+
+ENTRY = aws_cryptosdk_edk_list_copy_all_harness
+
+# The original aws_array_list_is_valid() has a 64 bit multiplication.
+# CBMC performance dies trying to do all those multiplications.
+# Replace with a stub until we can fix this issue.
+REMOVE_FUNCTION_BODIES += --remove-function-body aws_array_list_is_valid
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_edk_list_copy_all/Makefile:50]: Yes, I created one.

**Pull request merged**